### PR TITLE
Fix get_backend bug and add clear_device_cache function

### PR DESCRIPTION
**Pull request by NurmaU**
# What does this PR do?

This PR introduces 2 updates:
1. adds new `clear_device_cache` function in memory.py to replace repetitive code. This also adds check for mps devices in  function `find_executable_batch_size`. This feature was tested with tests/test_memory_utils.py, which had bug.
2. fixes bug in `get_backend` function where function's 3rd outputs must be callable object. This bug caused failures in tests/test_memory_utils.py on MacOS.

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/accelerate/test_utils/testing.py [hunk 1]
```diff
@@ -64,7 +64,7 @@ def get_backend():
     elif is_cuda_available():
         return "cuda", torch.cuda.device_count(), torch.cuda.memory_allocated
     elif is_mps_available(min_version="2.0"):
-        return "mps", 1, torch.mps.current_allocated_memory()
+        return "mps", 1, torch.mps.current_allocated_memory
```
Comment (BenjaminBossan) [inline src/accelerate/test_utils/testing.py:67]: Nice catch. But doesn't this mean that the return values below `"mps", 1, 0` (for old mps) and `"cpu", 1, 0` (for CPU) are also broken? Should they return `lambda: 0` instead?

Comment (muellerzr) [reply src/accelerate/test_utils/testing.py:67]: Agreed, we should use a `lambda` for those

Comment (NurmaU) [reply src/accelerate/test_utils/testing.py:67]: Yes, you're right. Added `lambda: 0`. Thanks for pointing that out!

**Review by muellerzr — APPROVED** [review]
Thanks for the refactor, and sorry for the delay in getting this in!

**Pull request merged**